### PR TITLE
Mapbox GL

### PR DESCRIPTION
**Pull request by atomiomi**
## API change

Marker props changed:
- `<Marker>tooltip` -> `<Marker>popupContent`
- `<Marker>content` -> `<Marker>children`

Popup was moved to marker:
- `<MarkerPopup>content` -> `<Marker>popupContent`
- `<MarkerPopup>options` -> `<Marker>popupOffset`
- `<MarkerPopup>defaultOpen` -> `<Marker>popupDefaultState`

## TODO:
- [x] Polygon, Circle sourceId generation
- [x] Polygon click handler
- [x] Map fitBounds breaks on update
- [x] Autofit map layers

### CODE REVIEW

File: src/map/index.es [hunk 1]
```diff
@@ -1,150 +1,42 @@
...
-      <article {...cleanProps} {...{ className, ref }}>
-        {content}
-      </article>
-    );
-  }
-}
+  return (
+    <MapboxMap
+      {...rest}
+      style="https://api.maptiler.com/maps/streets/style.json?key=h5gjGa1Ak2h0KgddSpXq"
```
Comment (max-degterev) [inline src/map/index.es]: get from args also

File: .gitignore [hunk 2]
```diff
@@ -35,4 +35,5 @@ node_modules/*
 lib/
 style.css
 preview/public/
+preview/config.es
```
Comment (max-degterev) [inline .gitignore]: use `uni-config` package and add default.js as an example

File: preview/containers/map/index.es [hunk 3]
```diff
@@ -104,31 +97,48 @@ class MapPreview extends PureComponent {
       mapLocked,
       mapLockedMobile,
       mapNoAttr,
-      mapAnimate,
```
Comment (max-degterev) [inline preview/containers/map/index.es]: animation?

File: package.json [hunk 4]
```diff
@@ -47,7 +47,7 @@
     "del": "^5.1.0",
     "errorhandler": "^1.5.1",
     "eslint": "^6.6.0",
-    "eslint-config-nebenan": "^1.3.1",
+    "eslint-config-nebenan": "^1.3.2",
     "express": "^4.17.1",
     "fibers": "^4.0.3",
     "gulp": "^4.0.2",
```
Comment (max-degterev) [inline package.json:53]: check if packages need updates/update

**Comment by max-degterev** [comment]
- [x] Locked map shows controls/allows zoom
- [x] Circle view isn't centered
- [x] Marker view isn't centered

File: preview/index.scss [hunk 5]
```diff
@@ -1,6 +1,6 @@
+@import 'mapbox-gl/dist/mapbox-gl';
```
Comment (peterhass) [inline preview/index.scss:1]: Does every user of this library have to import this by hand?
Suggestion: Create something like a index.scss file importing mapbox-gl so that users of nebenan-map don't need to know how it is implemented.

File: src/circle/index.es [hunk 6]
```diff
@@ -1,50 +1,27 @@
...
+import { getTypeProp, getPaint, getGeoJSON } from './utils';
+import { getID } from '../utils';
+
+
+const Circle = ({
+  type,
+  center,
+  radius,
+}) => {
+  const sourceId = useRef(getID()).current;
```
Comment (peterhass) [inline src/circle/index.es:16]: Nitpick: Would extract `sourceIdRef` - have never seen `useRef` used in a one-liner with `current`

File: src/circle/utils.es [hunk 7]
```diff
@@ -1,25 +1,40 @@
...
+    'circle-color': COLOR_BASE,
+    'circle-radius': 300,
   },
 };
 
-export const getOptions = ({ type, radius }) => {
-  const preset = STYLES[type];
-  return { ...preset, interactive: false, radius };
-};
+export const getTypeProp = () => PropTypes.oneOf([
```
Comment (peterhass) [inline src/circle/utils.es]: Do we need a function here? Maybe a variable would be sufficient too?

File: src/map/hooks.es [hunk 8]
```diff
@@ -0,0 +1,86 @@
+import { useEffect, useState, useRef } from 'react';
+import ReactMapboxGl from 'react-mapbox-gl';
+import { NavigationControl } from 'mapbox-gl';
+
+import { getMedia, media } from '../utils';
+import { getBoundsCenter, fitBounds } from './utils';
+
+
+export const useMapboxComponent = (locked, lockedMobile, noAttribution) => {
+  const [, setState] = useState();
```
Comment (peterhass) [inline src/map/hooks.es]: could it still work if you put ReactMapboxGl into state instead of a ref? That way you wouldn't have to manually force re-render

File: src/circle/index.es [hunk 9]
```diff
@@ -1,50 +1,27 @@
...
+  );
 };
 
 Circle.propTypes = {
-  type: PropTypes.oneOf([
-    CIRCLE_ACTIVE,
-    CIRCLE_DEFAULT,
-  ]).isRequired,
-
+  type: getTypeProp().isRequired,
```
Comment (max-degterev) [inline src/circle/index.es]: don't need a util here IMO, just spreads validation and can't be reused

File: src/circle_marker/index.es [hunk 10]
```diff
@@ -1,22 +1,24 @@
...
-
-  return <Marker {...cleanProps} divIcon={icon} />;
-};
+const CircleMarker = ({
+  children,
+  className,
+  ...rest
+}) => (
+  <Marker {...rest} className={clsx('c-circle_marker', className)}>
+    <span className="c-circle_marker-container">
```
Comment (max-degterev) [inline src/circle_marker/index.es:13]: Make children required?

File: src/eyecatcher_marker/index.es [hunk 11]
```diff
@@ -4,28 +4,21 @@ import clsx from 'clsx';
...
-  const html = (
-    `<div class="c-eyecatcher_marker-container">
-      ${content}
+const EyecatcherMarker = ({
+  children,
+  className,
+  ...rest
+}) => (
+  <Marker {...rest} className={clsx('c-eyecatcher_marker', className)}>
+    <div className="c-eyecatcher_marker-container">
```
Comment (max-degterev) [inline src/eyecatcher_marker/index.es:13]: Make children required?

File: src/label_marker/index.es [hunk 12]
```diff
@@ -4,24 +4,19 @@ import clsx from 'clsx';
...
+  className,
+  ...rest
+}) => (
+  <Marker {...rest} className={clsx('c-label_marker', className)}>
+    <span className="c-label_marker-container">{children}</span>
+  </Marker>
+);
 
 LabelMarker.propTypes = {
+  children: PropTypes.node,
```
Comment (max-degterev) [inline src/label_marker/index.es]: Make children required?

File: src/map/hooks.es [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+export const useContextValue = () => {
+  const boundsSet = useRef(new Set()).current;
+  const contextValue = useRef({
+    addBounds: (value) => {
+      boundsSet.add(value);
+      return () => boundsSet.delete(value);
+    },
+  }).current;
+
+  return [boundsSet, contextValue];
```
Comment (peterhass) [inline src/map/hooks.es]: would expect [get,set] for hooks returning array type. maybe object would be a bitter fit?

File: src/map/hooks.es (hunk 8 above)
Comment (max-degterev) [inline src/map/hooks.es]:
```suggestion
  const setState = useState()[1];
```

File: src/map/hooks.es [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+import { getMedia, media } from '../utils';
+import { getBoundsCenter, fitBounds } from './utils';
+
+
+export const useMapboxComponent = (locked, lockedMobile, noAttribution) => {
+  const [, setState] = useState();
+  const ref = useRef(null);
+
+  useEffect(() => {
+    const isMobile = !getMedia(global, media.mediaM);
```
Comment (max-degterev) [inline src/map/hooks.es:11]: didn't we already have mobile hook?

File: src/map/hooks.es [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+      scrollZoom: false,
+      injectCSS: false,
+      dragRotate: false,
+      pitchWithRotate: false,
+      apiUrl: null,
+      attributionControl: !noAttribution,
+    });
+
+    // Force re-render
+    setState({});
```
Comment (max-degterev) [inline src/map/hooks.es:36]: why not simply avoid rendering until calculations are complete? not render until ref.current is set or return isReady state?

File: src/map/hooks.es [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  }, [locked, lockedMobile, noAttribution]);
+
+  return ref.current;
+};
+
+export const useDefaultCenterAndZoom = (defaultZoom, defaultView, bounds) => {
+  const zoomRef = useRef(defaultZoom ? [defaultZoom] : undefined);
+  const centerRef = useRef((bounds ? getBoundsCenter(bounds) : defaultView) || undefined);
+
+  return [centerRef.current, zoomRef.current];
```
Comment (max-degterev) [inline src/map/hooks.es:46]: this doesn't seem right, you're centering on center and right boundary, not left boundary and right boundary. is this working properly?

do you need refs here at all? why not use a simple util?

File: src/utils/index.es [hunk 17]
```diff
@@ -1,9 +1,11 @@
...
-
 export const media = {
   mediaS: '(min-width: 450px)',
   mediaM: '(min-width: 690px)',
   mediaL: '(min-width: 920px)',
 };
 
 export const getMedia = (node, query) => node.matchMedia(query).matches;
+
+export const getID = () => (
```
Comment (max-degterev) [inline src/utils/index.es:9]: nebenan-helpers isnt imported?

File: src/polygon/index.es [hunk 18]
```diff
@@ -1,71 +1,35 @@
...
+const Polygon = (props) => {
+  const {
+    type,
+    area,
+    onClick,
+  } = props;
+
+  useAddBoundsToMap(area);
+
+  const sourceId = useRef(getID()).current;
```
Comment (max-degterev) [inline src/polygon/index.es:18]: create a hook?

File: src/marker/hooks.es [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+
+  return new Marker(element);
+};
+
+const createPopup = (win, content, options) => {
+  const html = renderToStaticMarkup(content);
+  return new Popup(options).setHTML(html);
+};
+
+export const useMarker = ({
```
Comment (max-degterev) [inline src/marker/hooks.es]: this isn't reused so just put the hook content back into component?

File: src/marker/hooks.es [hunk 20]
```diff
@@ -0,0 +1,51 @@
+import { useContext, useEffect } from 'react';
+import { renderToStaticMarkup } from 'react-dom/server';
```
Comment (max-degterev) [inline src/marker/hooks.es]: we shouldn't include this on the client

File: src/marker/hooks.es [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+  const element = win.document.createElement('div');
+  element.innerHTML = renderToStaticMarkup(content);
+  element.className = className;
+
+  return new Marker(element);
+};
+
+const createPopup = (win, content, options) => {
+  const html = renderToStaticMarkup(content);
+  return new Popup(options).setHTML(html);
```
Comment (max-degterev) [inline src/marker/hooks.es]: could render a child and then use setDOMContent to avoid strings

File: src/marker/hooks.es [hunk 22]
```diff
@@ -0,0 +1,51 @@
+import { useContext, useEffect } from 'react';
+import { renderToStaticMarkup } from 'react-dom/server';
+import { MapContext } from 'react-mapbox-gl';
+import { Marker, Popup } from 'mapbox-gl';
+
+
+const createMarker = (win, content, className) => {
+  const element = win.document.createElement('div');
+  element.innerHTML = renderToStaticMarkup(content);
```
Comment (max-degterev) [inline src/marker/hooks.es]: rendering children would work also

File: src/map/utils.es [hunk 23]
```diff
@@ -1,49 +1,29 @@
...
-  'tap',
-  'zoomControl',
-];
 
-const viewProps = [
-  'locked',
-  'lockedMobile',
-  'noAttribution',
-];
+export const getBoundsCenter = (bounds) => ([
```
Comment (max-degterev) [inline src/map/utils.es]: why do you need center? AFAIK possible to use bbox: [ bottom left,  top right ]

File: src/map/utils.es [hunk 24]
```diff
@@ -1,49 +1,29 @@
...
-
-  return { ...base, scrollWheelZoom: false, attributionControl: !noAttribution };
-};
-
-export const getTileUrl = (credentials) => (
-  credentials
-    ? `https://api.maptiler.com/maps/streets/256/{z}/{x}/{y}.png?key=${credentials.key}`
-    : 'http://{s}.tile.osm.org/{z}/{x}/{y}.png'
+export const getStyle = (credentials) => (
+  `https://api.maptiler.com/maps/075de9e7-0503-4e15-87af-7b99e074604c/style.json?key=${credentials.key}`
```
Comment (max-degterev) [inline src/map/utils.es]: don't hardcode IDs

File: src/map/utils.es [hunk 25]
```diff
@@ -1,49 +1,29 @@
...
-export const getTileOptions = (credentials) => {
-  const osm = '<a href="http://osm.org/copyright" target="_blank" rel="noopener noreferrer">&copy; OpenStreetMap contributors</a>';
-  if (!credentials) return { attribution: osm };
+export const getLngLatBounds = (bounds) => {
+  if (!Array.isArray(bounds) || !bounds.length) return undefined;
 
-  const attribution = `
-    <a href="https://www.maptiler.com/copyright/" target="_blank" rel="noopener noreferrer">&copy; MapTiler</a> ${osm}
-  `;
+  const initial = new LngLatBounds(bounds[0], bounds[0]);
```
Comment (max-degterev) [inline src/map/utils.es:14]: same coord twice?

File: src/map/index.scss [hunk 26]
```diff
@@ -1,22 +1,6 @@
 .c-map {
   @include block-responsive(75%);
-  position: relative;
 
   // overrides default leaflet font
   font-family: $font-family;
```
Comment (max-degterev) [inline src/map/index.scss:5]: still necessary? update comment

File: src/map/hooks.es (hunk 13 above)
Comment (max-degterev) [inline src/map/hooks.es]: .current chances, the wrapper persists. are you sure exporting the value works best?

File: src/map/hooks.es [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    if (callback) callback(map);
+  };
+
+  return [mapRef, loadHandler];
+};
+
+export const useBoundsUpdate = (mapRef, bounds) => {
+  useEffect(() => {
```
Comment (max-degterev) [inline src/map/hooks.es]: you should just use useEffect directly, this hook does almost nothing

File: src/map/hooks.es [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+
+  const loadHandler = (map) => {
+    mapRef.current = map;
+    map.addControl(new NavigationControl());
+
+    if (defaultBounds) {
+      fitBounds(map, defaultBounds, false);
+    } else {
+      // Wait for children render
+      process.nextTick(fitChildrenBounds);
```
Comment (max-degterev) [inline src/map/hooks.es]: why is this necessary, shouldn't children call `addBounds` after rendering is complete?

**Review by max-degterev — CHANGES_REQUESTED** [review]

File: src/map/hooks.es (hunk 8 above)
Comment (peterhass) [reply src/map/hooks.es]: I like `  const [, setState] = useState();` better.  Looks more familiar

File: src/map/utils.es [hunk 29]
```diff
@@ -1,49 +1,29 @@
...
+  const collect = (acc, item) => acc.extend(item);
 
-  return { attribution };
+  return bounds.reduce(collect, initial);
```
Comment (peterhass) [inline src/map/utils.es]: Nitpick: Would inline `collect` and `initial`. Call would be quite short anyway

File: src/polygon/hooks.es [hunk 30]
```diff
@@ -0,0 +1,8 @@
+import { useContext, useEffect } from 'react';
+import MapContext from '../map/context';
+
+
+export const useAddBoundsToMap = (area) => {
```
Comment (peterhass) [inline src/polygon/hooks.es]: `useMapBounds` ? 

`useAddBoundsToMap` sounds a bit like it would return a callback

File: src/map/hooks.es (hunk 8 above)
Comment (atomiomi) [reply src/map/hooks.es]: will leave this part as it is

File: src/map/hooks.es (hunk 14 above)
Comment (atomiomi) [reply src/map/hooks.es:11]: don't want to connect peer deps for one check only

File: src/map/hooks.es (hunk 15 above)
Comment (atomiomi) [reply src/map/hooks.es:36]: this is done to update component after ref is set.
With isReady state it will be rendered only one time after first initialization.
We need to be able to recreate this component after some props were changed

File: src/map/hooks.es (hunk 16 above)
Comment (atomiomi) [reply src/map/hooks.es:46]: react-mapbox-gl updates view after each center, zoom props update.
Using ref to set these props only on first render.

will check center point

File: src/utils/index.es (hunk 17 above)
Comment (atomiomi) [reply src/utils/index.es:9]: nope

File: src/polygon/index.es (hunk 18 above)
Comment (atomiomi) [reply src/polygon/index.es:18]: very simple hook that is used only in 2 places. not worth it

File: src/map/utils.es (hunk 25 above)
Comment (atomiomi) [reply src/map/utils.es:14]: yep. got this part from mapbox docs https://docs.mapbox.com/mapbox-gl-js/example/zoomto-linestring/

File: preview/index.scss (hunk 5 above)
Comment (atomiomi) [reply preview/index.scss:1]: react-mapbox -gl imports this styles in js. Wouldn't need this on new build (goodhood repo)

File: src/circle/index.es (hunk 6 above)
Comment (atomiomi) [reply src/circle/index.es:16]: using ref to call getID only one time. Won't be changed no need to keep whole ref

File: src/map/hooks.es (hunk 8 above)
Comment (atomiomi) [reply src/map/hooks.es]: react breaks if i put component in state

File: src/map/hooks.es (hunk 13 above)
Comment (atomiomi) [reply src/map/hooks.es]: return array is better to renaming variables

File: src/map/index.scss (hunk 26 above)
Comment (atomiomi) [reply src/map/index.scss:5]: yep, mapbox uses helvetica by default

File: package.json (hunk 4 above)
Comment (atomiomi) [reply package.json:53]: will update after moving to goodhood

File: src/map/utils.es (hunk 25 above)
Comment (max-degterev) [reply src/map/utils.es:14]: leave a comment

File: src/marker/utils.es [hunk 31]
```diff
@@ -1,4 +1,17 @@
...
-  if (divIcon) return leaflet.divIcon(divIcon);
-  if (defaultIcon) return leaflet.icon(defaultIcon);
+import { render } from 'react-dom';
+import { Marker, Popup } from 'mapbox-gl';
+
+
+export const createMarker = (win, content, className) => {
+  const element = win.document.createElement('div');
+  element.className = className;
+  render(content, element);
```
Comment (max-degterev) [inline src/marker/utils.es:7]: why not just render a node within marker and then pass a ref to it here?

**Review by max-degterev — APPROVED** [review]

**Comment by max-degterev** [comment]
- [x] Locked map - reset cursor to default (now hand)

File: src/map/index.scss (hunk 26 above)
Comment (max-degterev) [reply src/map/index.scss:5]: update comment
